webgl: out-of-bounds readPixels() fixes

anholt opened the pull request:
<!-- Please describe your changes on the following line: -->
Fix crashes in two WebGL readPixels() tests by adding framebuffer size validation.

---
<!-- Thank you for contributing to Servo! Please replace each `[ ]` by `[X]` when the step is complete, and replace `__` with appropriate data: -->
- [X] `./mach build -d` does not report any errors
- [X] `./mach test-tidy` does not report any errors
- [X] These changes fix #13901

<!-- Either: -->
- [X] There are tests for these changes OR
- [ ] These changes do not require tests because _____

<!-- Pull requests that do not address these steps are welcome, but they will require additional verification as part of the review process. -->

<!-- Reviewable:start -->

---
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/servo/servo/14081)
<!-- Reviewable:end -->

anholt committed:
webgl: Throw an error on readPixels(width < 0 || height < 0)

Otherwise gleam will try to allocate a negative size area for the
result, and we'll panic with oom.

anholt committed:
webgl: Fix out-of-bounds readpixels handling.

This fixes the crash in read-pixels-pack-alignment (which was trying
to read out of bounds).

Fixes #13901

anholt committed:
webgl: Add an interface for getting the size of renderbuffers.

We need this for framebuffer status checking.

anholt committed:
webgl: Track the level with texture attachments.

We need this to be able to get the size of the attached texture for
completeness validation.

Signed-off-by: Eric Anholt <eric@anholt.net>

anholt committed:
webgl: Validate that framebuffer attachment sizes match.

This is required by the WebGL spec, and we need to figure out the FB
size like this for validating ReadPixels.

anholt committed:
webgl: Only update FBO status once in framebufferTexture2D().

We update it below after we send the actual FramebufferTexture2D
command to the GL.

anholt committed:
webgl: Don't forget to update FBO status after detaching a texture.

We had this in the renderbuffer detach process, but missed it in this
one.

anholt committed:
webgl: Update FBO status when textures or RBs are reallocated.

FBO status is supposed to depend on the size of the attachments all
matching, so we need to re-check when it changes.  We don't ensure
matching yet, but this will prevent regressions when we do.

highfive commented:
Heads up! This PR modifies the following files:
 * @kichjang: components/script/dom/webglframebuffer.rs, components/script/dom/webgltexture.rs, components/script/dom/webglrenderingcontext.rs, components/script/dom/webglrenderbuffer.rs
 * @fitzgen: components/script/dom/webglframebuffer.rs, components/script/dom/webgltexture.rs, components/script/dom/webglrenderingcontext.rs, components/script/dom/webglrenderbuffer.rs
 * @emilio: components/script/dom/webglframebuffer.rs, components/script/dom/webgltexture.rs, components/script/dom/webglrenderingcontext.rs, components/script/dom/webglrenderbuffer.rs

emilio commented on components/script/dom/webglframebuffer.rs:
```diff
@@ -21,7 +21,7 @@ use webrender_traits::{WebGLCommand, WebGLFramebufferBindingRequest, WebGLFrameb
 enum WebGLFramebufferAttachment {
     Renderbuffer(JS<WebGLRenderbuffer>),
-    Texture(JS<WebGLTexture>),
+    Texture(JS<WebGLTexture>, i32),
```
A note saying this is the texture level would be nice. Alternatively:

`Texture { texture: JS<WebGLTexture>, level: i32 }`

anholt replied:
Oh, nice.  I thought the JS<> in there meant the struct would be harder than just doing that.

emilio commented on components/script/dom/webglframebuffer.rs:
```diff
@@ -190,7 +190,7 @@ impl WebGLFramebuffer {
             //      "If texture is zero, then textarget and level are ignored."
             Some(texture) => {
-                *binding.borrow_mut() = Some(WebGLFramebufferAttachment::Texture(JS::from_ref(texture)));
+                *binding.borrow_mut() = Some(WebGLFramebufferAttachment::Texture(JS::from_ref(texture), level));
```
should we move this after the level validation? Otherwise in your next patch we can't be convert it to an unsigned integer right away.

emilio commented on components/script/dom/webglframebuffer.rs:
```diff
@@ -135,6 +140,32 @@ impl WebGLFramebuffer {
+                Some(WebGLFramebufferAttachment::Renderbuffer(ref att_rb)) => {
+                    att_rb.size()
+                }
+                Some(WebGLFramebufferAttachment::Texture(ref att_tex, ref level)) => {
```
nit: Binding level by value is probably ok

emilio commented on components/script/dom/webglframebuffer.rs:
```diff
@@ -251,7 +251,8 @@ impl WebGLFramebuffer {
 
-    pub fn detach_renderbuffer(&self, rb: &WebGLRenderbuffer) {
+    fn call_on_matching_renderbuffers<F>(&self, rb: &WebGLRenderbuffer, closure: F)
+        where F: Fn(&DOMRefCell<Option<WebGLFramebufferAttachment>>) {
```
nit: naming convention for this kind of function is `with`, like `with_matching_{renderbuffers/textures}`.

Also, when there's a `where` clause usually the opening brace goes on the next line aligned with the `fn`:

```rust
fn with_matching_renderbuffers<F>(&self, rb: &WebGLRenderbuffer, closure: F)
    where F: Fn(&DOMRefCell<Option<WebGLFramebufferAttachment>>)
{
    // ...
```

emilio replied:
Also, you can probably use `FnMut` instead of `Fn`, which would allow you to use closures that mutate other state.

emilio commented on components/script/dom/webglframebuffer.rs:
```diff
@@ -267,13 +268,13 @@ impl WebGLFramebuffer {
 
-    pub fn detach_texture(&self, texture: &WebGLTexture) {
+    fn call_on_matching_textures<F>(&self, texture: &WebGLTexture, closure: F)
+        where F: Fn(&DOMRefCell<Option<WebGLFramebufferAttachment>>) {
```
same comment as above.

emilio commented on components/script/dom/webglframebuffer.rs:
```diff
@@ -289,12 +290,37 @@ impl WebGLFramebuffer {
+    pub fn detach_renderbuffer(&self, rb: &WebGLRenderbuffer) {
+        self.call_on_matching_renderbuffers(rb, |att| {
+            *att.borrow_mut() = None;
+            self.update_status();
```
I think with the `FnMut` clause we can track here and below whether at least one has been matched and update the status only once?

anholt replied:
Neat idea.  I think I'll stick with the simple version here, since you couldn't have a complete FBO with the same object attached twice anyway.

emilio reviewed (approved):
r=me with those. Looks great, thanks!

anholt committed:
webgl: Don't update texture attachment state if we throw an error.

If we're going to return from this function without updating the
underlying GL state because of a WebGLError, we shouldn't have updated
our shadow of the GL state.

emilio commented:
@bors-servo try

bors-servo commented:
:hourglass: Trying commit f32b6da with merge eb2ef4f...

emilio commented:
@bors-servo r+

Bors forgot about this apparently, huh

bors-servo commented:
:pushpin: Commit eaec5de has been approved by `emilio`

<!-- @bors-servo r=emilio eaec5de1849beb6f7b3e464851044c9429a23537 -->

bors-servo commented:
:hourglass: Testing commit eaec5de with merge 2601d8e...

bors-servo commented:
:broken_heart: Test failed - [mac-rel-wpt1](http://build.servo.org/builders/mac-rel-wpt1/builds/763)

jdm commented:
@bors-servo: retry
* #14152

bors-servo commented:
:zap: Previous build results for [arm32](http://build.servo.org/builders/arm32/builds/3999), [arm64](http://build.servo.org/builders/arm64/builds/4005), [linux-dev](http://build.servo.org/builders/linux-dev/builds/4098), [linux-rel-css](http://build.servo.org/builders/linux-rel-css/builds/741), [linux-rel-wpt](http://build.servo.org/builders/linux-rel-wpt/builds/740), [mac-dev-unit](http://build.servo.org/builders/mac-dev-unit/builds/4083), [mac-rel-css](http://build.servo.org/builders/mac-rel-css/builds/4079), [mac-rel-wpt2](http://build.servo.org/builders/mac-rel-wpt2/builds/776), [windows-dev](http://build.servo.org/builders/windows-dev/builds/2011) are reusable. Rebuilding only [mac-rel-wpt1](http://build.servo.org/builders/mac-rel-wpt1/builds/763)...

bors-servo commented:
:sunny: Test successful - [arm32](http://build.servo.org/builders/arm32/builds/3999), [arm64](http://build.servo.org/builders/arm64/builds/4005), [linux-dev](http://build.servo.org/builders/linux-dev/builds/4098), [linux-rel-css](http://build.servo.org/builders/linux-rel-css/builds/741), [linux-rel-wpt](http://build.servo.org/builders/linux-rel-wpt/builds/740), [mac-dev-unit](http://build.servo.org/builders/mac-dev-unit/builds/4083), [mac-rel-css](http://build.servo.org/builders/mac-rel-css/builds/4079), [mac-rel-wpt1](http://build.servo.org/builders/mac-rel-wpt1/builds/766), [mac-rel-wpt2](http://build.servo.org/builders/mac-rel-wpt2/builds/776), [windows-dev](http://build.servo.org/builders/windows-dev/builds/2011)
